query-engine/core/src/query_graph/mod.rs

```diff
@@ -412,6 +412,18 @@ impl QueryGraph {
         })
     }
 
+    /// Returns all leaf nodes of the graph.
+    /// A leaf node is defined by having no outgoing edges.
+    pub fn leaf_nodes(&self) -> impl Iterator<Item = NodeRef> + '_ {
+        self.graph.node_indices().filter_map(|node_ix| {
+            self.graph
+                .edges_directed(node_ix, Direction::Outgoing)
+                .next()
+                .is_none()
+                .then_some(NodeRef { node_ix })
```

Comment:
It could technically be just a `map` but I like how nicely this reads like a sentence so 🤷